src/mono/browser/runtime/cancelable-promise.ts

```diff
@@ -32,17 +38,121 @@ export function wrap_as_cancelable<T>(inner: Promise<T>): ControllablePromise<T>
 export function mono_wasm_cancel_promise(task_holder_gc_handle: GCHandle): void {
     const holder = _lookup_js_owned_object(task_holder_gc_handle) as PromiseHolder;
     mono_assert(!!holder, () => `Expected Promise for GCHandle ${task_holder_gc_handle}`);
+    holder.cancel();
+}
 
-    const promise = holder.promise;
-    loaderHelpers.assertIsControllablePromise(promise);
-    const promise_control = loaderHelpers.getPromiseController(promise);
-    if (holder.isResolved) {
-        // FIXME: something needs to free the GCHandle
-        mono_log_warn("Canceling a promise that has already resolved.");
-        return;
-    }
-    mono_assert(!holder.isCanceled, "This promise already canceled.");
-    holder.isCanceled = true;
-    promise_control.reject(new Error("OperationCanceledException"));
+// NOTE: layout has to match PromiseHolderState in JSHostImplementation.Types.cs
+const enum PromiseHolderState {
+    IsResolving = 0,
 }
 
+const promise_holder_symbol = Symbol.for("wasm promise_holder");
+
+export class PromiseHolder extends ManagedObject {
+    public isResolved = false;
+    public isPosted = false;
+    public data: any = null;
+    public reason: any = null;
+    public constructor(public promise: Promise<any>,
+        private gc_handle: GCHandle,
+        private promiseHolderPtr: number, // could be null for GCV_handle 
+        private res_converter?: MarshalerToCs) {
+        super();
+    }
+
+    // returns true if the promise is being canceled by managed code
```

Comment:
if I understand this code right, a better comment would be:
```
// attempts to transition the promise into a 'being resolved' state.
// returns true if the state was transitioned successfully and you can now resolve the promise.
// returns false if the state was not transitioned, and it is unsafe for you to resolve the promise.

Reply:
actually, the boolean returns seem inverted? which is confusing to me. i think it might be better if success was true and failure was false.